mobile/src/main/java/be/mygod/vpnhotspot/tasker/TetheringActions.kt

```diff
@@ -0,0 +1,108 @@
+
+import android.content.Context
+import android.os.Build
+import android.os.Bundle
+import androidx.annotation.RequiresApi
+import androidx.appcompat.app.AppCompatActivity
+import be.mygod.vpnhotspot.net.TetheringManager
+import com.joaomgcd.taskerpluginlibrary.action.TaskerPluginRunnerActionNoOutput
+import com.joaomgcd.taskerpluginlibrary.config.TaskerPluginConfig
+import com.joaomgcd.taskerpluginlibrary.config.TaskerPluginConfigHelperNoOutput
+import com.joaomgcd.taskerpluginlibrary.input.TaskerInput
+import com.joaomgcd.taskerpluginlibrary.input.TaskerInputField
+import com.joaomgcd.taskerpluginlibrary.input.TaskerInputRoot
+import com.joaomgcd.taskerpluginlibrary.runner.TaskerPluginResult
+import com.joaomgcd.taskerpluginlibrary.runner.TaskerPluginResultError
+import com.joaomgcd.taskerpluginlibrary.runner.TaskerPluginResultSucess
+import kotlinx.coroutines.runBlocking
+import kotlin.coroutines.resume
+import kotlin.coroutines.suspendCoroutine
+
+@TaskerInputRoot
+data class ActionInput(
+    @TaskerInputField("tetherType")
+    var tetherType: Int = -1,
+    @TaskerInputField("enable")
+    var enable: Boolean = false,
+)
+
+sealed class ActionConfig(tetherType: Int) : AppCompatActivity(), TaskerPluginConfig<ActionInput> {
+    override val context by lazy { this }
+    override val inputForTasker: TaskerInput<ActionInput>
+        get() = TaskerInput(input)
+
+    private val helper by lazy { ActionHelper(this) }
+
+    private var input: ActionInput = ActionInput(
+        enable = false,
+        tetherType = tetherType,
+    )
+
+    override fun assignFromInput(input: TaskerInput<ActionInput>) {
+        this.input = this.input.copy(enable = !input.regular.enable)
+    }
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+
+        helper.onCreate()
+        helper.finishForTasker()
+    }
+
+    class WifiConfig : ActionConfig(TetheringManager.TETHERING_WIFI)
+    class USBConfig : ActionConfig(TetheringManager.TETHERING_USB)
+    class BluetoothConfig : ActionConfig(TetheringManager.TETHERING_BLUETOOTH)
+    @RequiresApi(Build.VERSION_CODES.R)
+    class EthernetConfig : ActionConfig(TetheringManager.TETHERING_ETHERNET)
+}
+
+class ActionHelper(
+    config: TaskerPluginConfig<ActionInput>,
+) : TaskerPluginConfigHelperNoOutput<ActionInput, ActionRunner>(config) {
+    override val inputClass: Class<ActionInput> = ActionInput::class.java
+    override val runnerClass: Class<ActionRunner> = ActionRunner::class.java
+    override val addDefaultStringBlurb: Boolean = false
+
+    override fun addToStringBlurb(input: TaskerInput<ActionInput>, blurbBuilder: StringBuilder) {
+        blurbBuilder.appendLine("enable: ${input.regular.enable}")
+    }
+}
+
+class ActionRunner : TaskerPluginRunnerActionNoOutput<ActionInput>() {
+    override fun run(context: Context, input: TaskerInput<ActionInput>): TaskerPluginResult<Unit> {
+        return runBlocking {
```

Comment:
Hi thanks again. It seems to work but this is not great. There is no reason to use a thread pool to run essentially nothing. Is there a more lightweight implementation?

Reply:
This is how João demonstrates running async tasks inside an action: https://github.com/joaomgcd/TaskerPluginSample/blob/5e16316f66516008abbb86431f29dba6c494cc8e/app/src/main/java/com/joaomgcd/taskerpluginsample/tasker/backgroundwork/BackgroundWork.kt#L67.

It doesn't look like the Android API has a synchronous way to invoke tethering operations, and Tasker has no callback ability, so this is the only way to return a result from the action.

All Tasker actions also already run inside a thread pool, and I think `runBlocking` inherits its parent thread when you don't specify a dispatcher.

Reply:
Okay I guess this library is not that great but that's not your fault. I will probably merge this but won't maintain these code. Feel free to contribute further in the future. :)